source/beamline/AMTemperatureMonitorGroupStatus.cpp

```diff
@@ -0,0 +1,158 @@
+#include "AMTemperatureMonitorGroupStatus.h"
+#include "beamline/AMTemperatureMonitor.h"
+
+AMTemperatureMonitorGroupStatus::AMTemperatureMonitorGroupStatus(const QString &name, QObject *parent) :
+	AMEnumeratedControl(name, "", parent)
+{
+	// Setup value options.
+
+	addOption(Bad, "Bad", true);
+	addOption(Good, "Good", true);
+}
+
+AMTemperatureMonitorGroupStatus::~AMTemperatureMonitorGroupStatus()
+{
+
+}
+
+bool AMTemperatureMonitorGroupStatus::canMeasure() const
+{
+	bool result = false;
+
+	if (isConnected()) {
+
+		// Iterate through the list of monitors, finding out if
+		// all can be measured.
+
+		QList<AMControl*> monitors = childControls();
+
+		if (monitors.count() > 0) {
+			bool measurable = true;
+
+			for (int i = 0, count = monitors.count(); i < count && measurable; i++) { // we want to stop if any one child is not measurable.
+				AMTemperatureMonitor *monitor = qobject_cast<AMTemperatureMonitor*>(monitors.at(i));
+
+				if ( !(monitor && monitor->canMeasure()) )
+					measurable = false;
+			}
+
+			result = measurable;
+		}
+	}
+
+	return result;
+}
+
+bool AMTemperatureMonitorGroupStatus::isBad() const
+{
+	bool monitorFound = false;
+
+	// Iterate through list of monitors, finding out if
+	// any are in the bad state.
+
+	QList<AMControl*> monitors = childControls();
+
+	if (monitors.count() > 0) {
+
+		for (int i = 0, count = monitors.count(); i < count && !monitorFound; i++) { // we want to stop if any one child is in a bad state.
+			AMTemperatureMonitor *monitor = qobject_cast<AMTemperatureMonitor*>(monitors.at(i));
+
+			if (monitor && monitor->isBad())
+				monitorFound = true;
+		}
+	}
+
+	return monitorFound;
+}
+
+bool AMTemperatureMonitorGroupStatus::isGood() const
+{
+	bool badMonitorFound = false;
+
+	// Iterate through list of monitors, finding out if
+	// all are in a good state.
+
+	QList<AMControl*> monitors = childControls();
+
+	if (monitors.count() > 0) {
+
+		for (int i = 0, count = monitors.count(); i < count && !badMonitorFound; i++) { // we want to stop if any one child isn't in state 1.
+			AMTemperatureMonitor *monitor = qobject_cast<AMTemperatureMonitor*>(monitors.at(i));
+
+			if (monitor && !monitor->isGood())
```

Comment:
I think that while the interface being presented is essentially boolean (i.e.: Good or Bad) that one should have the necessary logic and then the other be the negation of that.  The way that you implemented isBad, I think isGood could just be the negation.

Conversely, if you had implemented isGood such that every monitor needed to be good, then isBad() would just be !isGood().